feat: write objects to blob storage

rjsparks opened the pull request.

rjsparks committed:
feat: basic blobstore infrastructure for dev

rjsparks committed:
refactor: (broken) attempt to put minio console behind nginx

rjsparks committed:
feat: initialize blobstore with boto3

rjsparks committed:
fix: abandon attempt to proxy minio. Use docker compose instead.

rjsparks committed:
ci: Merge branch 'main' into feat/blob

rjsparks committed:
Merge branch 'main' into feat/blob

rjsparks committed:
feat: beginning of blob writes

rjsparks committed:
feat: storage utilities

rjsparks committed:
feat: test buckets

rjsparks committed:
chore: black

rjsparks committed:
chore: remove unused import

rjsparks committed:
chore: avoid f string when not needed

rjsparks committed:
fix: inform all settings files about blobstores

rjsparks committed:
fix: declare types for some settings

rjsparks committed:
ci: point to new target base

rjsparks committed:
Merge remote-tracking branch 'ietf-tools/main' into feat/blobwrites

rjsparks committed:
ci: adjust test workflow

rjsparks committed:
fix: give the tests debug environment a blobstore

rjsparks committed:
fix: "better" name declarations

rjsparks committed:
Merge remote-tracking branch 'ietf-tools/main' into feat/blobwrites

rjsparks committed:
ci: use devblobstore container

rjsparks committed:
chore: identify places to write to blobstorage

rjsparks committed:
Merge remote-tracking branch 'ietf-tools/main' into feat/blobwrites

rjsparks committed:
chore: remove unreachable code

rjsparks committed:
Merge remote-tracking branch 'rjsparks/noglobunlink' into feat/blobwrites

rjsparks committed:
feat: store materials

rjsparks committed:
feat: store statements

rjsparks committed:
feat: store status changes

rjsparks committed:
feat: store liaison attachments

rjsparks committed:
feat: store agendas provided with Interim session requests

rjsparks committed:
chore: capture TODOs

rjsparks committed:
feat: store polls and chatlogs

rjsparks committed:
chore: remove unneeded TODO

rjsparks committed:
Merge remote-tracking branch 'ietf-tools/main' into feat/blobwrites

rjsparks committed:
feat: store drafts on submit and post

rjsparks committed:
fix: handle storage during doc expiration and resurrection

jennifer-richards commented on ietf/doc/expire.py:
```diff
@@ -225,6 +225,13 @@ def move_file_to(subdir):
+                # See, however, the note about "major system failures" at "unknown_ids"
+                blobname = f"{ext[1:]}/{basename}"
+                if exists_in_storage("active-draft", blobname):
+                    remove_from_storage("active-draft", blobname)
```
Just call `remove_from_storage()` and handle an exception if it complains about blobs that don't exist? The code needs to handle that possibility anyway, e.g., due to a race.

rjsparks replied:
The implemenation in django-storages (and in boto3) explicitly say it is not an error to remove something that doesn't exist.

jennifer-richards reviewed (commented):
Sorry if this a comment on code not intended to be ready for review, but I happened to look... one comment inline. 

Also, for testing I wonder if we want a fake blob Storage class that does only what we need for that it's being called as intended. If we make that work with your `storage_utils`, the tests wouldn't even have to know. If the need arose, we could also test against a real blob store by adjusting the configuration.

rjsparks committed:
fix: mirror an unlink

rjsparks committed:
chore: add/refine TODOs

rjsparks committed:
feat: store slide submissions

jennifer-richards committed:
feat: BlobShadowFileSystemStorage

rjsparks committed:
fix: structure slide test correctly

rjsparks committed:
fix: correct sense of existence check

rjsparks committed:
feat: store some indexes

jennifer-richards committed:
feat: shadow floorplans / host logos to the blob

jennifer-richards committed:
chore: remove unused import

jennifer-richards committed:
feat: strip path from blob shadow names

jennifer-richards committed:
Merge remote-tracking branch 'upstream/feat/blobwrites' into imageblob

# Conflicts:
#	ietf/settings.py

jennifer-richards committed:
feat: shadow photos / thumbs

jennifer-richards committed:
refactor: combine photo and photothumb blob kinds

The photos / thumbs were already dropped in the same
directory, so let's not add a distinction at this point.

jennifer-richards committed:
style: whitespace

jennifer-richards committed:
refactor: use kwargs consistently

jennifer-richards committed:
chore: migrations

jennifer-richards committed:
refactor: better deconstruct(); rebuild migrations

rjsparks committed:
feat: shadow image storage to blob store (#8477)

rjsparks committed:
fix: use new class in mack patch

jennifer-richards committed:
chore: add TODO

rjsparks committed:
feat: store group index documents

rjsparks committed:
chore: identify more TODO

rjsparks committed:
feat: store reviews

rjsparks committed:
Merge remote-tracking branch 'ietf-tools/main' into feat/blobwrites

rjsparks committed:
Merge remote-tracking branch 'ietf-tools/main' into feat/blobwrites

rjsparks committed:
fix: repair merge

rjsparks committed:
chore: remove unnecessary TODO

rjsparks committed:
feat: StoredObject metadata

rjsparks committed:
fix: deburr some debugging code

jennifer-richards commented on ietf/doc/storage_backends.py:
```diff
@@ -0,0 +1,137 @@
+                new_name = self.save(name, file)
+                now = timezone.now()
+                existing_record = StoredObject.objects.filter(store=kind, name=name)
+                if existing_record.exists():
```
Should use `update_or_create()`

jennifer-richards commented on ietf/doc/storage_backends.py:
```diff
@@ -0,0 +1,137 @@
+                    complaint = f"Error encountered saving '{name}' - results stored in '{new_name}' instead."
+                    log(complaint)
+                    debug.show("complaint")
+                    # Note that we are otherwise ignoring this condition - it should become an error later.
```
Might as well raise an exception now since we have the log-and-swallow in place

rjsparks replied:
leaving this wart for now - need to think through what it should raise when we take away log-and-swallow.

jennifer-richards commented on ietf/doc/storage_backends.py:
```diff
@@ -0,0 +1,137 @@
+                del self.in_flight_custom_metadata[name]
+        return None
+
+    def exists_in_storage(self, kind: str, name: str) -> bool:
```
again, why not override?

rjsparks replied:
again, parallel structure. Also I think we want to (for the short term) keep the parent `.exists()` bare outside this overridden class just for continued learning.

jennifer-richards commented on ietf/doc/storage_backends.py:
```diff
@@ -0,0 +1,137 @@
+        except FileNotFoundError:
+            return False
+
+    def remove_from_storage(
```
... and again ... override the remove method?

jennifer-richards commented on ietf/doc/storage_backends.py:
```diff
@@ -0,0 +1,137 @@
+        except FileNotFoundError:
+            if warn_if_missing:
+                complaint = (
+                    f"WARNING: Asked to delete non-existant {name} from {kind} storage"
```
spelling: exist**e**nt

jennifer-richards commented on ietf/doc/storage_backends.py:
```diff
@@ -0,0 +1,137 @@
+        params = super()._get_write_parameters(name, content)
+        if "Metadata" not in params:
+            params["Metadata"] = {}
+        if not is_seekable(content):
```
`is_seekable` is [a strange method](https://github.com/jschneier/django-storages/blob/b79ea310201e7afd659fe47e2882fe59aae5b517/storages/utils.py#L122) - it returns `True` if the `content` does not have a `seekable()` method, so e.g. `is_seekable("strings don't seek")` is `True`. Wrapping the `content.seek(0)` in try/except would be a better guard than using the helper.

jennifer-richards commented on ietf/doc/storage_backends.py:
```diff
@@ -0,0 +1,137 @@
+            debug.show("complaint")
+        else:
+            # Note that existing_record is a queryset that will have one matching object
+            existing_record.update(deleted=now)
```
This probably ought to add a `.filter(deleted__isnull=True)` to the queryset for the `update()`. That will keep `deleted` as the _earliest_ timestamp if a file is deleted more than once.

(We don't want the deleted filter for the `warn_if_missing` existence check, though)

jennifer-richards reviewed (commented):
Posting as comment to share these while I continue reviewing

jennifer-richards commented on ietf/doc/models.py:
```diff
@@ -1538,3 +1560,31 @@ class BofreqEditorDocEvent(DocEvent):
+        help_text="Last instant object was modified. May not be the same as the storage's modified value for the instance. It will hold mtime for objects imported from older disk storage unless they've actually been overwritten more recently"
+    )
+    doc_name = models.CharField(max_length=255, null=True, blank=True)
+    doc_rev = models.CharField(max_length=16, null=True, blank=True)
```
This is not necessariiy a request for a change, but I'm having misgivings about this quasi-ForeignKey link to `Document`. We've talked about it a bit, but I wanted to write it down because it came to mind again. I think it's ok to proceed with the name/rev fields initially, but I just get a tingling that it's going to turn into another wart on the already hard-to-work-with Document class hierarchy that we'll be complaining about later.

Would this be better implemented as a many-to-many between `Document` and `StoredObject`? We landed here because `Document` gets mixed up in `DocHistory`, but there could _also_ be a many-to-many that links `DocHistory` and `StoredObject` that is populated when a `Document` becomes a `DocHistory`. The usual way we save docs should handle this for us, since it has logic to deal with many-to-many fields.

This is a little bit unpleasant because it's really a one-to-many between `Document` and `StoredObject`, and then a many-to-many between `DocHistory` and `StoredObject`. If that really bothers us, we could handle that by customizing the handling along the lines we do for `DocumentAuthor`.

Are there objections to doing this beyond the complexity involved in copying the many-to-many field around?

rjsparks commented on ietf/doc/storage_backends.py:
```diff
@@ -0,0 +1,137 @@
+        self.in_flight_custom_metadata: Dict[str, Dict[str, str]] = {}
+        return super().__init__(**settings)
+
+    def store_file(
```
Mostly to keep the structure parallel for ease of readability. But you trigger several thoughts with the question.

* I think it would be a loss to let the general datatracker code call the storage's save directly - nothing outside the utility managing the storages should know it's actually storages underneath. Specifically, I want to avoid the assumption that there's only _one_ storage or storage strategy underneath.
* We may _have_ to override save here anyhow if we want things that use the *Field implementations which _do_ expect to be bare on top of a Storage to get this metadata handling.

rjsparks committed:
fix: only set the deleted timestamp once

rjsparks committed:
chore: correct typo

rjsparks committed:
fix: get_or_create vs get and test

rjsparks committed:
fix: avoid the questionable is_seekable helper

rjsparks committed:
chore: capture future design consideration
